src/ui/state_manager.cc

```diff
@@ -0,0 +1,50 @@
+#include "state_manager.hh"
+#include <QTimer>
+#include <QMutexLocker>
+
+StateManager::StateManager( QWidget * parent, Config::Class & cfg ):
+  _cfg( cfg )
+{
+  _state    = cfg.mainWindowState;
+  _geometry = cfg.mainWindowGeometry;
+  _dirty    = false;
+}
+
+
+void StateManager::setState( QByteArray state, QByteArray geometry )
+{
+  if ( _dirty )
+    return;
+
+
+  if ( state == _state && _geometry == geometry ) {
+    return;
+  }
+  _dirty = true;
+
+  //delay execution
+  QTimer::singleShot( 3000, this, [ = ]() {
+    saveConfigData( state, geometry );
+
+    _dirty = false;
+  } );
+}
+
+
+void StateManager::saveConfigData( QByteArray state, QByteArray geometry )
+{
+  QMutexLocker locker( &_mutex );
```

Comment:
This mutex doesn't seem to be protecting anything. The functions invoked by QTimer::singleShot happens in the main Thread.

If for extra protection, the `config::save` should have a mutex instead?